### ISSUE TITLE
[USD-231] Bug(frontend): posts list display problem

### ISSUE THREAD

**Issue by kittybest** [social-tw/social-tw-website#520]
**Describe the bug**
[description of the bug here](https://github.com/social-tw/social-tw-website/pull/519#issuecomment-2381974389)

Also if user refresh the page, the latest posts are not shown.

**To Reproduce**

1. sign up as user 1
2. sign up as user 2
3. wait until the 1st epoch passes, the 2nd epoch starts
4. user 1 do user state transition and make a post
5. switch back to user 2's post list

Another steps;
1. sign up
2. wait until the 1st epoch passes
3. make a post
4. refresh the page

**Expected behavior**
everyone should see the same post list

- `PostService::updateOrder`
1. should add an argument to setup the query page of the post table. the statement should also be modified to add `LIMIT` and `OFFSET` clause to fetch the given page of posts https://github.com/social-tw/social-tw-website/blob/9cc927d0b2cebc0d550c3463265bc8dc16a0ee05/packages/relay/src/services/PostService.ts#L35
2. should return the posts directly without saving into the cache. https://github.com/social-tw/social-tw-website/blob/9cc927d0b2cebc0d550c3463265bc8dc16a0ee05/packages/relay/src/services/PostService.ts#L116-L122
3. remove the instance which invoke updateOrder method.
  a. `reportRoute::/api/report`
  b. `UnirepSocialSynchronizer::handleEpochEnded`
- `PostService::fetchPosts`
1. should replace the cache with updateOrder https://github.com/social-tw/social-tw-website/blob/9cc927d0b2cebc0d550c3463265bc8dc16a0ee05/packages/relay/src/services/PostService.ts#L151-L167
- `PostService::start`
1. after remove the cache mechanism, the start function will no longer be invoked. should remove the start method and the `cache` property.
2. remove the postService.start(db) on index.ts https://github.com/social-tw/social-tw-website/blob/9cc927d0b2cebc0d550c3463265bc8dc16a0ee05/packages/relay/src/index.ts#L41
**Screenshots**
in the link above

**Comment by chentihe** [comment]
current post list is using cache, so the order of posts will be updated every 30 mins, to solve the post list display issue, we should disable the cache so that post list from every users screen would be the latest.

**Comment by cc03668** [comment]
@chentihe 請協助編輯這一則 Issue 附上我們討論過後的解方

**Comment by Lisooo790926** [comment]
HI @kittybest @chentihe 
My understanding requirement is 
1. Remove all cache
2. Put update order into `fetchPosts`

pros : 
1. Simple implemtation 
cons : 
1. Hard SQL maintainability
2. Query loading is too high if we need to update the order every fetching

Maybe we could add some redundant columns to the Post table like the `score`, `dailyCommentCount`, `dailyUpCount` etc., we could hourly recalculate these columns, and query the post sorted by `score`
To achieve it, 
1. Table schema should add 
  ```javascript
  export type Post = {
      postId: string | undefined
      content: string | undefined
      publishedAt: string
      transactionHash: string | undefined
      cid: string | undefined
      epoch: number
      epochKey: string
      upCount: number
      downCount: number
      voteSum: number
      status: PostStatus
      commentCount: number
      
      // add here 
      score: number
      dailyUpCount: number
      dailyDownCount: number
      dailyCommentCount: number
    
      votes: any[]
      _id: string
  }
  ```
2. [`Scheduler`]We should have a Scheduler to update post tables hourly, we could follow below (just examples )
  ```sql
  -- get postId -> comment count
  select count(1), postId from comment group by postId where publishedAt = now() - 24 hour; 
  
  -- get postId -> up vote count 
  select count(1), postId from vote group by postId where upVote = 1 and publishedAt = now() - 24 hour; 
  
  -- get postId -> down vote count 
  select count(1), postId from vote group by postId where downVote = 1 and publishedAt = now() - 24 hour; 
  ```
3. [`Scheduler`]Bulk update above counts into post table
  ``` javascript
  const bulkOperations = counts.map(({ postId, dailyCount}) => ({
    updateOne: {
      filter: { postId: postId },
      update: { $set: { dailyCommentCount: dailyCount.comment, dailyDownCount: dailyCount.downCount, dailyDownCount: dailyCount.upCount } }
    }
  }));
  ```
4. [`Scheduler`]Recalculate the score of each post in sql (the post created in 2 days, 1000 is for additional score to make the latest post reveal)
  ```sql
  update post set score = 
      DATEDIFF(NOW(), publishedAt) * (-0.5) + upCount * 0.2 -  downCount * 0.2 + commentCount * 0.1 + 1000
  where publishedAt >= now() - 48 hour; 
  ```
5. [`Scheduler`]Recalculate the score of each post in sql (the post created more than 2 days)
  ```sql
  update post set score = dailyUpCount + dailyCommentCount
  where publishedAt < now() - 48 hour; 
  ```
6. [`PostService`]fetch posts could be like below 
  ```sql
  select * from post sorted by score desc limit pageSize offset pageSize * (currentPage - 1);
  ```

pros: 
1. Query DB efficiency (although there is no cache)
2. SQL is easily maintainable 
3. Compared to the original method, we don't need to update the order in each fetch, we move this loading into the scheduler (hourly update makes server loading small)
cons : 
Need more time to implementation 

 Just raise what I see and discuss, If you want me to implement by following the original way, it's fine. 

Thanks

**Comment by Lisooo790926** [comment]
I think step 4, 5, 6 could be simpler
new 4 -> 4 + 5
new 5 -> 4 + 6

since only more than two days case need to update daily columns

**Comment by Lisooo790926** [comment]
Should check with others first since the sorting logic is already set. 
Sorting by creation time would make things a lot easier if that's an option

**Comment by ryanycw** [comment]
I was confused about this issue, I thought we were adding the logic of querying. Since we already have the query command, why don't we just use the same query logic and query without maintaining the cache? Is it a time concern?

**Comment by ryanycw** [comment]
I was confused about the discussion. Going through the context again, I think we can just move forward and implement with the description in this issue, if no one wants to add to the discussion. 

I feel like the new approach will add too much complexity to this issue.

**Comment by ryanycw** [comment]
Closed by #577
